src/optimize.rs

```diff
@@ -0,0 +1,79 @@
+use crate::metadata::*;
+use crate::utils::*;
+use tempfile::NamedTempFile;
+use walrus::*;
+use wasm_opt::OptimizationOptions;
+
+pub fn optimize(m: &mut Module, keep_name_section: bool, level: &str) {
+    // recursively optimize embedded modules in Motoko actor classes
+    if is_motoko_canister(m) {
+        let data = get_motoko_wasm_data_sections(m);
+        for (id, mut module) in data.into_iter() {
+            optimize(&mut module, keep_name_section, level);
+            let blob = encode_module_as_data_section(module);
+            m.data.get_mut(id).value = blob;
+        }
+    }
+
+    // pull out a copy of the custom sections to preserve
+    let m_copy = parse_wasm(&m.emit_wasm(), keep_name_section).unwrap();
+    let mut metadata_sections = Vec::new();
+    list_metadata(&m_copy).iter().for_each(|full_name| {
+        match full_name.strip_prefix("icp:public ") {
+            Some(name) => {
+                metadata_sections.push(("public", name, get_metadata(&m_copy, name).unwrap()))
+            }
+            None => match full_name.strip_prefix("icp:private ") {
+                Some(name) => {
+                    metadata_sections.push(("private", name, get_metadata(&m_copy, name).unwrap()))
+                }
+                None => unreachable!(),
+            },
+        };
+    });
```

Comment:
This is equivalent to the following,

```suggestion
    let metadata_sections = list_metadata(m);
```

Reply:
I believe `list_metadata` does not retrieve the actual sections, only the names.